src/main/java/org/opensearch/neuralsearch/query/HybridQueryDocIdStream.java

```diff
@@ -44,41 +44,32 @@ public int count(int upTo) {
+        // Always get the current matching bitset from the bulk scorer
+        FixedBitSet matchingBitSet = hybridBulkScorer.getMatching();
         long[] bitArray = matchingBitSet.getBits();
-        // iterate through each block of 64 documents (since each long contains 64 bits)
+
         for (int idx = 0; idx < bitArray.length; idx++) {
             long bits = bitArray[idx];
             while (bits != 0L) {
-                // find position of the rightmost set bit (1)
                 int numberOfTrailingZeros = Long.numberOfTrailingZeros(bits);
-                // calculate actual document ID within the window
-                // idx << 6 is equivalent to idx * 64 (block offset)
-                // numberOfTrailingZeros gives position within the block
                 final int docIndexInWindow = (idx << BLOCK_SHIFT) | numberOfTrailingZeros;
                 final int docId = base | docIndexInWindow;
 
-                // Only process documents up to the specified limit
-                if (docId >= upTo) {
-                    return;
-                }
-
-                float[][] windowScores = hybridBulkScorer.getWindowScores();
-                for (int subQueryIndex = 0; subQueryIndex < windowScores.length; subQueryIndex++) {
-                    if (Objects.isNull(windowScores[subQueryIndex])) {
-                        continue;
+                // Only process documents < upTo
+                if (docId < upTo) {
+                    float[][] windowScores = hybridBulkScorer.getWindowScores();
+                    for (int subQueryIndex = 0; subQueryIndex < windowScores.length; subQueryIndex++) {
+                        if (Objects.isNull(windowScores[subQueryIndex])) {
+                            continue;
+                        }
+                        float scoreOfDocIdForSubQuery = windowScores[subQueryIndex][docIndexInWindow];
+                        hybridBulkScorer.getHybridSubQueryScorer().getSubQueryScores()[subQueryIndex] = scoreOfDocIdForSubQuery;
                     }
-                    float scoreOfDocIdForSubQuery = windowScores[subQueryIndex][docIndexInWindow];
-                    hybridBulkScorer.getHybridSubQueryScorer().getSubQueryScores()[subQueryIndex] = scoreOfDocIdForSubQuery;
+                    consumer.accept(docId);
+                    hybridBulkScorer.getHybridSubQueryScorer().resetScores();
                 }
-                // process the document with its base offset
-                consumer.accept(docId);
-                // reset scores after processing of one doc, this is required because scorer object is re-used
-                hybridBulkScorer.getHybridSubQueryScorer().resetScores();
-                // clear bit from the local bitset copy to indicate that it has been consumed
-                matchingBitSet.clear(docIndexInWindow);
-                // reset bit for this doc id to indicate that it has been consumed
+
+                // Don't clear bits from the shared bitset!
```

Comment:
Since clearing the bits didn't let us process the next set of documents, should we have a local and global copy of bitset? Local will be the current window and global would be for all rather than working on just one

Reply:
I think having a global copy will have few drawbacks:
- extra resource consumption, currently resources are freed after each window has been processed, with global state we have to maintain it for duration of the query
- concurrency issues like race conditions and synchronization overhead, if we start synchronize on bitset state we loose the benefit of parallelism and thus the gain in response times
I do like a mental model of a streaming video, you don't load entire clip into memory, but you process it frame by frame, where each frame is independent.